outlines/text/generate/regex.py

```diff
@@ -0,0 +1,160 @@
+import math
+from typing import List, Optional, Tuple
+
+import interegular
+import torch
+
+from outlines.text.generate.continuation import Continuation
+from outlines.text.parsing import find_partial_matches, map_partial_states_to_vocab
+
+
+class Regex(Continuation):
+    """Represents a regex-based generation model.
+
+    `Regex` instances are constrained generation models that only generate
+    sequences that match an input regex. We assume that the sequence can be
+    terminated (but not necessarily) when the finite state machine corresponding
+    to the regex is in an accepting state.
+
+    >>> import outlines.text as text
+    >>> sequence = text.generate.regex(model, "(0|[1-9][0-9]+)")("Return an integer between 0 and 10")
+
+    """
+
+    def __init__(self, model, regex_string: str, max_tokens: Optional[int]):
+        super().__init__(model, max_tokens)
+
+        vocabulary = model.tokenizer.vocabulary
+        sorted_vocabulary = [
+            k for k, v in sorted(vocabulary.items(), key=lambda kv: kv[1])
+        ]
+
+        regex_pattern = interegular.parse_pattern(regex_string)
+        self.regex_fsm = regex_pattern.to_fsm().reduce()
+
+        def partial_match_filter(string, end_idx, state_seq):
+            if end_idx is not None and end_idx < len(string) - 1:
+                return False
+            return True
+
+        pstate_to_vocab = map_partial_states_to_vocab(
+            list(sorted_vocabulary),
+            {"REGEX": self.regex_fsm},
+            partial_match_filter,
+            final_state_string=model.tokenizer.eos_token,
+        )
+
+        # TODO: This check might be a little too strict, because I think that
+        # while some states are made unreachable by a vocabulary (and will not
+        # be present in the following set difference), there could still be
+        # paths to terminal states emanating from the states that are reachable.
+        states_with_transition = {x[1] for x in pstate_to_vocab.keys()}
+        if len(self.regex_fsm.states.difference(states_with_transition)) > 0:
+            raise ValueError(
+                "The vocabulary does not allow us to build a sequence that matches the input regex"
+            )
```

Comment:
We should look into this.  Perhaps as a follow-up issue.

Reply:
Good point. I opened #184 to track this.